cbor/shared/src/main/scala/fs2/data/cbor/high/internal/ValueParser.scala

```diff
@@ -0,0 +1,321 @@
+              }
+            }
+          }
+      }
+    }
+
+  private def parseByteStrings[F[_]](chunk: Chunk[CborItem],
+                                     idx: Int,
+                                     rest: Stream[F, CborItem],
+                                     acc: ByteVector,
+                                     chunkAcc: List[CborValue])(implicit
+      F: RaiseThrowable[F]): Pull[F, CborValue, Result[F, CborValue]] =
+    if (idx >= chunk.size) {
+      Pull.output(Chunk.seq(chunkAcc.reverse)) >> rest.pull.uncons.flatMap {
+        case Some((hd, tl)) => parseByteStrings(hd, 0, tl, acc, Nil)
+        case None           => Pull.raiseError(new CborException("unexpected end of input"))
+      }
+    } else {
+      chunk(idx) match {
+        case CborItem.Break =>
+          Pull.pure((chunk, idx + 1, rest, chunkAcc, CborValue.ByteString(acc)))
+        case CborItem.ByteString(bytes) =>
+          if (acc.size + bytes.size < 0l) {
+            raise(new CborException(s"byte string size is limited to max long (${Long.MaxValue}) bits"), chunkAcc)
+          } else {
+            parseByteStrings(chunk, idx + 1, rest, acc ++ bytes, chunkAcc)
+          }
+        case item =>
+          raise(new CborException(s"unexpected item $item"), chunkAcc)
+      }
+    }
+
+  private def parseTextStrings[F[_]](chunk: Chunk[CborItem],
+                                     idx: Int,
+                                     rest: Stream[F, CborItem],
+                                     acc: StringBuilder,
+                                     chunkAcc: List[CborValue])(implicit
+      F: RaiseThrowable[F]): Pull[F, CborValue, Result[F, CborValue]] =
+    if (idx >= chunk.size) {
+      Pull.output(Chunk.seq(chunkAcc.reverse)) >> rest.pull.uncons.flatMap {
+        case Some((hd, tl)) => parseTextStrings(hd, 0, tl, acc, Nil)
+        case None           => Pull.raiseError(new CborException("unexpected end of input"))
+      }
+    } else {
+      chunk(idx) match {
+        case CborItem.Break =>
+          Pull.pure((chunk, idx + 1, rest, chunkAcc, CborValue.TextString(acc.result())))
+        case CborItem.TextString(text) =>
+          if (acc.size + text.size < 0) {
+            raise(new CborException(s"text string size is limited to max int (${Int.MaxValue}) characters"), chunkAcc)
+          } else {
+            parseTextStrings(chunk, idx + 1, rest, acc.append(text), chunkAcc)
+          }
+        case item =>
+          raise(new CborException(s"unexpected item $item"), chunkAcc)
+      }
+    }
+
+  private def parseTags[F[_]](chunk: Chunk[CborItem],
+                              idx: Int,
+                              rest: Stream[F, CborItem],
+                              tags: CborValue => CborValue,
+                              chunkAcc: List[CborValue])(implicit
+      F: RaiseThrowable[F]): Pull[F, CborValue, Result[F, CborValue => CborValue]] =
+    if (idx >= chunk.size) {
```

Comment:
I always wonder whether this could be extract to an `@inline` method to keep the code DRY.

Reply:
I thought about it and didn't find anything satisfying so far.

Reply:
Fancy way: Use `ContT`. But it feels like it could also be done within the monadic context of `Pull`, so things become more `flatMap`py and can maybe even be written as for-comprehension.

Reply:
Trying out something like this:

```scala
  def ensureChunk[F[_], T](chunk: Chunk[Byte], idx: Int, rest: Stream[F, Byte], chunkAcc: List[CborItem])(
      cont: (Chunk[Byte], Int, Stream[F, Byte], List[CborItem]) => Pull[F, CborItem, T])(
      onEos: => Pull[F, CborItem, T]): Pull[F, CborItem, T] =
    if (idx >= chunk.size) {
      Pull.output(Chunk.seq(chunkAcc)) >> rest.pull.uncons.flatMap {
        case Some((hd, tl)) => cont(hd, 0, tl, Nil)
        case None           => onEos
      }
    } else {
      cont(chunk, idx, rest, chunkAcc)
    }
```

I will need to tune the compiler settings to inline it, though

Reply:
I pushed something, if you could have a look at it it would be great.